delegate-app/Main.hs

```diff
@@ -1,49 +1,179 @@
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (
+    DelegateServerConfig,
+    dlgt'host,
+    dlgt'ping,
+    dlgt'port,
+    dlgt'tick
+  ),
+  DelegateServerLog (
+    DelegateError,
+    DelegateOutput,
+    FrontendConnected,
+    FrontendInput,
+    Started
+  ),
+  DelegateTracerT,
+  ServerAppT,
+ )
+import HydraAuction.Delegate.Tracing (MonadTracer (trace), TracerT, askTracer, runWithTracer')
+
+{- | per websocket, a thread gets spawned which then
+   subscribes to the broadcast channel and sends to
+   the incoming channel of the delegate runner
+-}
+delegateServerApp ::
+  -- | the amount of seconds between pings
+  Int ->
+  -- | the queue for the inputs of the delegates (this is where the requests are enqueued)
+  TQueue DelegateInput ->
+  -- | the queue for the broadcast (this is where we receives events from the runner)
+  TChan DelegateResponse ->
+  ServerAppT (DelegateTracerT IO)
+delegateServerApp pingSecs reqq broadcast pending = do
+  connection <- liftIO $ acceptRequest pending
+
+  trace FrontendConnected
+  tracer <- askTracer
+
+  liftIO $ do
+    broadcast' <- atomically $ dupTChan broadcast
+
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
+        encodeSend = sendTextData connection . encode
+
+        receive = forever $
+          runWithTracer' tracer $ do
+            inp <- liftIO $ receiveData connection
+            case eitherDecode @FrontendRequest inp of
+              Left msg -> do
+                let err = FrontendNoParse msg
+                trace (DelegateError err)
+                liftIO $ encodeSend err
+              Right req -> do
+                trace $ FrontendInput req
+                liftIO . atomically . writeTQueue reqq $ FrontendRequest req
+
+        send =
+          forever $
+            atomically (readTChan broadcast')
+              >>= encodeSend
+
+    withPingThread connection pingSecs (pure ()) $ race_ receive send -- TODO: can we do this
 
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
-eventsProducer thisIsFirstRun =
-  if thisIsFirstRun
-    then return $ Just Start
-    else return Nothing
-
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
-
-delegateTick :: Int
-delegateTick = 1_000
+{- | create a delegate runner that will listen for incoming events and output outgoing events
+   accordingly
+-}
+mkRunner ::
+  -- | the time in milliseconds that the runner sleeps between acts
+  Int ->
+  -- | the queue of incoming messages
+  TQueue DelegateInput ->
+  -- | the broadcast queue of outgoing messages (write only)
+  TChan DelegateResponse ->
+  DelegateRunnerT (DelegateTracerT IO) ()
```

Comment:
I think that term `Runner` is already used as name for monad pattern we use, not for actual code block.

Reply:
I think I don't understand this.

Reply:
We use word `Runner` to name monad pattern. You name kind of thread the same term. But it is different meaning for same term - we use `Runner` monad(s) in multiple different situations. That is such term selection is confusing.